# Admissions Classifier

Binary classifier that predicts **Accepted** vs **Rejected** for undergraduate college applications using GPA, SAT score, and extracurricular activities.

**Project requirements (instructor)**: University/College Admissions — a pipeline from beginning to end. Beginning: a student comes to the university/college for an admission enquiry. End product: Admitted.

**Objective**: Train a neural network on the [Kaggle student admission dataset](https://www.kaggle.com/datasets/amanace/student-admission-dataset).

**Model type**: Binary classification (sigmoid output, binary cross-entropy).

**Workflow**:
1. Load data
2. Preprocess and binarize target (Accepted=1, Rejected/Waitlisted=0)
3. Train/test split, scale features
4. Build and train neural network
5. Evaluate and save model + scaler

## Run in the browser (no local setup)

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/adzuci/admissions-classifier/blob/main/admissions_classifier.ipynb)

Works on **Colab**, **Windows**, and **Mac**. Colab fetches the dataset from GitHub. Local Mac: `pip install tensorflow-macos tensorflow-metal`. Local Windows: `pip install tensorflow`. Ensure `student_admission_dataset.csv` is in the folder if running locally (download from [Kaggle](https://www.kaggle.com/datasets/amanace/student-admission-dataset) if needed).

In [25]:
# Colab: install tensorflow, sklearn, etc. Local: pip install tensorflow-macos tensorflow-metal scikit-learn
try:
    import google.colab
    get_ipython().run_line_magic('pip', 'install -q tensorflow pandas joblib scikit-learn')
except (ImportError, NameError):
    pass  # local: deps assumed installed (Mac: tensorflow-macos+tensorflow-metal; Windows: tensorflow)

## STEP 1: Imports and SimpleScaler

In [26]:
# Mac: tensorflow-macos + tensorflow-metal (Metal GPU). Colab & Windows: standard tensorflow.
import joblib
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from tensorflow import keras
from tensorflow.keras import layers


class SimpleScaler:
    """Minimal scaler with fit/transform (no sklearn)."""

    def fit(self, X):
        self.mean_ = np.mean(X, axis=0)
        self.scale_ = np.std(X, axis=0)
        self.scale_[self.scale_ == 0] = 1.0
        return self

    def transform(self, X):
        return (X - self.mean_) / self.scale_


print(f"TensorFlow: {keras.__version__}")

TensorFlow: 3.12.1


## STEP 2: Load data

In [27]:
# Colab: fetch from GitHub. Local (Mac/Windows): use student_admission_dataset.csv
try:
    import google.colab
    csv_name = "https://raw.githubusercontent.com/adzuci/admissions-classifier/main/student_admission_dataset.csv"
except ImportError:
    csv_name = "student_admission_dataset.csv"

df = pd.read_csv(csv_name)
df.head(10)

GPA  SAT_Score  Extracurricular_Activities Admission_Status
0  3.46       1223                           8         Rejected
1  2.54        974                           8         Rejected
2  2.91        909                           9         Rejected
3  2.83       1369                           5         Accepted
4  3.60       1536                           7         Accepted
5  3.52       1476                           9         Accepted
6  3.84       1002                           8         Rejected
7  2.63        975                           6       Waitlisted
8  3.13       1450                           8       Waitlisted
9  2.54       1118                           7         Rejected

## STEP 3: Preprocess

In [28]:
# Binarize: Accepted=1, Rejected/Waitlisted=0
df["Admit"] = (df["Admission_Status"] == "Accepted").astype(int)

feature_cols = ["GPA", "SAT_Score", "Extracurricular_Activities"]
X = df[feature_cols].values.astype(np.float32)
y = df["Admit"].values

print(f"X shape: {X.shape}, y shape: {y.shape}")
print(f"Class balance: Admit=1 {y.sum()}, Admit=0 {len(y) - y.sum()}")

X shape: (250, 3), y shape: (250,)
Class balance: Admit=1 81, Admit=0 169


In [29]:
# Stratified train/test split (80/20) — preserves class balance
np.random.seed(42)
idx_1 = np.where(y == 1)[0]
idx_0 = np.where(y == 0)[0]
np.random.shuffle(idx_1)
np.random.shuffle(idx_0)
n_test_1 = max(1, int(len(idx_1) * 0.2))
n_test_0 = max(1, int(len(idx_0) * 0.2))
test_idx = np.concatenate([idx_1[:n_test_1], idx_0[:n_test_0]])
train_idx = np.concatenate([idx_1[n_test_1:], idx_0[n_test_0:]])
np.random.shuffle(train_idx)
np.random.shuffle(test_idx)

X_train, X_test = X[train_idx], X[test_idx]
y_train, y_test = y[train_idx], y[test_idx]

# Scale features
scaler = SimpleScaler()
scaler.fit(X_train)
X_train_s = scaler.transform(X_train)
X_test_s = scaler.transform(X_test)

## STEP 4: Build and train model

In [30]:
model = keras.Sequential([
    keras.Input(shape=(3,)),
    layers.Dense(32, activation="relu"),
    layers.Dropout(0.2),
    layers.Dense(16, activation="relu"),
    layers.Dropout(0.2),
    layers.Dense(1, activation="sigmoid"),
])
model.compile(optimizer=keras.optimizers.Adam(learning_rate=0.001), loss="binary_crossentropy", metrics=["accuracy"])
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_9 (Dense)                 │ (None, 32)             │           128 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_6 (Dropout)             │ (None, 32)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_7 (Dropout)             │ (None, 16)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_11 (Dense)                │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 673 (2.63 KB)

 Trainable params: 673 (2.63 KB)

 Non-trainable params: 0 (0.00 B)

In [31]:
# Class weights to handle imbalance (81 Accepted vs 169 Rejected/Waitlisted)
n_pos, n_neg = int(y_train.sum()), len(y_train) - int(y_train.sum())
class_weight = {0: 1.0, 1: n_neg / n_pos}

early_stop = keras.callbacks.EarlyStopping(
    monitor="val_loss", patience=15, restore_best_weights=True, verbose=0
)
history = model.fit(
    X_train_s, y_train,
    epochs=100, batch_size=16,
    validation_split=0.15,
    class_weight=class_weight,
    callbacks=[early_stop],
    verbose=1,
)

Epoch 1/100
11/11 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.5765 - loss: 0.9682 - val_accuracy: 0.3871 - val_loss: 0.7423
Epoch 2/100
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.5059 - loss: 0.9677 - val_accuracy: 0.3548 - val_loss: 0.7407
Epoch 3/100
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.5471 - loss: 0.9668 - val_accuracy: 0.3548 - val_loss: 0.7377
Epoch 4/100
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.5412 - loss: 0.9587 - val_accuracy: 0.3871 - val_loss: 0.7385
Epoch 5/100
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.5412 - loss: 0.9260 - val_accuracy: 0.4516 - val_loss: 0.7347
Epoch 6/100
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.5294 - loss: 0.9511 - val_accuracy: 0.4839 - val_loss: 0.7404
Epoch 7/100
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.5529 - loss: 0.9499 - val_accuracy: 0.4839 - val_loss: 0.7397
Epoch 8/100
11/11 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.5471 - loss: 0.9704 - val_accuracy: 0.3

## STEP 5: Evaluate

In [32]:
loss, acc = model.evaluate(X_test_s, y_test, verbose=0)
print(f"Test accuracy: {acc:.4f}")

preds = (model.predict(X_test_s, verbose=0) > 0.5).astype(int).flatten()
print(f"Sample predictions: {preds[:15]}...")

Test accuracy: 0.5102
Sample predictions: [1 0 1 0 1 1 0 0 1 0 0 1 0 0 1]...


**Why is accuracy ~59%?** Small dataset (250 samples, ~50 test) and only 3 features. Many students with similar GPA/SAT/extracurriculars have different outcomes (accepted vs rejected), so there's inherent ambiguity. Neural nets tend to need more data; simpler models often do better on small tabular data.

## Alternative models: Logistic Regression & Random Forest

Compare with classical ML models that often outperform neural nets on small tabular data.

In [33]:
# Train and evaluate Logistic Regression and Random Forest
lr = LogisticRegression(max_iter=500, class_weight="balanced", random_state=42)
rf = RandomForestClassifier(n_estimators=100, class_weight="balanced", random_state=42)

lr.fit(X_train_s, y_train)
rf.fit(X_train_s, y_train)

acc_nn = model.evaluate(X_test_s, y_test, verbose=0)[1]
acc_lr = (lr.predict(X_test_s) == y_test).mean()
acc_rf = (rf.predict(X_test_s) == y_test).mean()

print("Test accuracy comparison:")
print(f"  Neural net:    {acc_nn:.4f}")
print(f"  Logistic Reg: {acc_lr:.4f}")
print(f"  Random Forest: {acc_rf:.4f}")
print(f"\nBest: {'Logistic Regression' if acc_lr >= max(acc_nn, acc_rf) else 'Random Forest' if acc_rf >= max(acc_nn, acc_lr) else 'Neural net'}")

Test accuracy comparison:
  Neural net:    0.5102
  Logistic Reg: 0.6531
  Random Forest: 0.6327

Best: Logistic Regression


## Synthetic dataset: Do NNs do better with more data?

Generate a larger synthetic dataset (500 samples) with a **clear weighted rule** for admission. On small, noisy real data, simple models often win; on larger data with a learnable pattern, neural nets may do better.

In [34]:
# Create synthetic dataset (GPA 2–4, SAT 800–1600, activities 0–10)
# Weighted rule: GPA×25 (~50–100), SAT×0.05 (~40–80), activities×3 (~0–30)
np.random.seed(42)
n_synth = 500
hs_gpa_input = np.random.uniform(2.0, 4.0, n_synth)
sat_scores_input = np.random.randint(800, 1601, n_synth)
extra_activities_input = np.random.randint(0, 11, n_synth)

score = (hs_gpa_input * 25) + (sat_scores_input * 0.05) + (extra_activities_input * 3)
threshold = np.percentile(score, q=50)  # top 50% get admitted
noise = np.random.normal(loc=0, scale=3, size=n_synth)  # introduce some unpredictability
y_synth = (score + noise >= threshold).astype(np.int32)
X_synth = np.column_stack([hs_gpa_input, sat_scores_input, extra_activities_input]).astype(np.float32)

print(f"Synthetic: X shape {X_synth.shape}, Admit=1 {y_synth.sum()}, Admit=0 {n_synth - y_synth.sum()}")

Synthetic: X shape (500, 3), Admit=1 242, Admit=0 258


In [35]:
# Split, scale, train NN/LR/RF on synthetic data
np.random.seed(42)
idx1 = np.where(y_synth == 1)[0]
idx0 = np.where(y_synth == 0)[0]
np.random.shuffle(idx1)
np.random.shuffle(idx0)
n_test_1 = max(1, int(len(idx1) * 0.2))
n_test_0 = max(1, int(len(idx0) * 0.2))
test_idx = np.concatenate([idx1[:n_test_1], idx0[:n_test_0]])
train_idx = np.concatenate([idx1[n_test_1:], idx0[n_test_0:]])
np.random.shuffle(train_idx)
np.random.shuffle(test_idx)

X_tr, X_te = X_synth[train_idx], X_synth[test_idx]
y_tr, y_te = y_synth[train_idx], y_synth[test_idx]

scaler_synth = SimpleScaler()
scaler_synth.fit(X_tr)
X_tr_s = scaler_synth.transform(X_tr)
X_te_s = scaler_synth.transform(X_te)

# NN
model_synth = keras.Sequential([
    keras.Input(shape=(3,)),
    layers.Dense(32, activation="relu"),
    layers.Dropout(0.2),
    layers.Dense(16, activation="relu"),
    layers.Dropout(0.2),
    layers.Dense(1, activation="sigmoid"),
])
model_synth.compile(optimizer=keras.optimizers.Adam(learning_rate=0.001), loss="binary_crossentropy", metrics=["accuracy"])
n_pos, n_neg = int(y_tr.sum()), len(y_tr) - int(y_tr.sum())
cw = {0: 1.0, 1: n_neg / n_pos}
model_synth.fit(X_tr_s, y_tr, epochs=80, batch_size=32, validation_split=0.15, class_weight=cw, callbacks=[keras.callbacks.EarlyStopping(monitor="val_loss", patience=10, restore_best_weights=True, verbose=0)], verbose=0)

# LR & RF
lr_synth = LogisticRegression(max_iter=500, class_weight="balanced", random_state=42)
rf_synth = RandomForestClassifier(n_estimators=100, class_weight="balanced", random_state=42)
lr_synth.fit(X_tr_s, y_tr)
rf_synth.fit(X_tr_s, y_tr)

acc_nn_s = model_synth.evaluate(X_te_s, y_te, verbose=0)[1]
acc_lr_s = (lr_synth.predict(X_te_s) == y_te).mean()
acc_rf_s = (rf_synth.predict(X_te_s) == y_te).mean()

print("Synthetic dataset — test accuracy:")
print(f"  Neural net:      {acc_nn_s:.4f}")
print(f"  Logistic Reg:    {acc_lr_s:.4f}")
print(f"  Random Forest:   {acc_rf_s:.4f}")
best = max([("Neural net", acc_nn_s), ("Logistic Regression", acc_lr_s), ("Random Forest", acc_rf_s)], key=lambda x: x[1])
print(f"\nBest: {best[0]}")

Synthetic dataset — test accuracy:
  Neural net:      0.9495
  Logistic Reg:    0.9394
  Random Forest:   0.9293

Best: Neural net


## STEP 6: Save model and scaler

In [36]:
model.save("model.keras")
joblib.dump(scaler, "scaler.joblib")
print("Saved model.keras and scaler.joblib")
print("In Colab: use Files pane to download these files.")

Saved model.keras and scaler.joblib
In Colab: use Files pane to download these files.


## Predict a single student

In [37]:
def predict_student_admission(gpa, sat_score, extracurricular_activities):
    student = np.array([[gpa, sat_score, extracurricular_activities]])
    student_s = scaler.transform(student)

    prob_nn = model.predict(student_s, verbose=0)[0][0]
    pred_nn = "ACCEPTED" if prob_nn > 0.5 else "REJECTED"
    pred_lr = "ACCEPTED" if lr.predict(student_s)[0] else "REJECTED"
    pred_rf = "ACCEPTED" if rf.predict(student_s)[0] else "REJECTED"

    print(f"Student: GPA={student[0,0]}, SAT={int(student[0,1])}, Extracurriculars={int(student[0,2])}")
    print(f"  Neural net:      {pred_nn} (prob={prob_nn:.3f})")
    print(f"  Logistic Reg:   {pred_lr}")
    print(f"  Random Forest:  {pred_rf}")

In [38]:
# Run prediction
predict_student_admission(3.5, 1400, 7)

Student: GPA=3.5, SAT=1400, Extracurriculars=7
  Neural net:      ACCEPTED (prob=0.526)
  Logistic Reg:   ACCEPTED
  Random Forest:  REJECTED
